### PR TITLE
Discovery Server TSAN fixes [16106]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

## Description

This PR is meant to palliate issues using Discovery Server related with TSAN.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Fast DDS test suite has been run locally. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by jparisu** [comment]
So far this PR only implements a new class `Atomicable` for Elbrus team to decide whether they would like to use it or not.

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 1]
```diff
@@ -159,7 +159,6 @@ void DomainParticipantImpl::disable()
     {
         participant->set_listener(nullptr);
     }
-    rtps_listener_.participant_ = nullptr;
```
Comment (juanlofer-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:162]: Please remove the first commit from this PR.

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 2]
```diff
@@ -1590,6 +1593,8 @@ fastrtps::rtps::CacheChange_t* DiscoveryDataBase::cache_change_own_participant()
 
 const std::vector<fastrtps::rtps::GuidPrefix_t> DiscoveryDataBase::direct_clients_and_servers()
 {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
```
Comment (juanlofer-eprosima) [inline src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:1596]: NIT: use lock_guard when sufficient.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/database/DiscoverySharedInfo.hpp [hunk 3]
```diff
@@ -20,6 +20,8 @@
 #ifndef _FASTDDS_RTPS_DISCOVERY_SHARED_INFO_H_
 #define _FASTDDS_RTPS_DISCOVERY_SHARED_INFO_H_
 
+#include <mutex>
```
Comment (juanlofer-eprosima) [inline src/cpp/rtps/builtin/discovery/database/DiscoverySharedInfo.hpp:23]:
```suggestion
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by jsantiago-eProsima** [comment]
@richiprosima please test this

**Comment by jsantiago-eProsima** [comment]
@richiprosima please test linux

File: src/cpp/rtps/builtin/discovery/database/DiscoverySharedInfo.hpp [hunk 4]
```diff
@@ -102,7 +102,7 @@ class DiscoverySharedInfo
     virtual void to_json(
             nlohmann::json& j) const;
 
-private:
+protected:
```
Comment (jsantiago-eProsima) [inline src/cpp/rtps/builtin/discovery/database/DiscoverySharedInfo.hpp:105]: Why is this changed to protected?

Comment (jparisu) [reply src/cpp/rtps/builtin/discovery/database/DiscoverySharedInfo.hpp:105]: It made sense before when there was a mutex in this class, so children could use it. Now it is not needed.
However, as it is a class to be inherited and the common implementation along the project is to use protected, I think it is better to leave it this way

**Review by jsantiago-eProsima — APPROVED** [review]

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by MiguelCompany** [comment]
CI for 2.9.0-rc:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_linux/3187)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_mac/2944)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_windows/3349/)
* Discovery server [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/3131/)
